# MIDTERM

In [1]:
import pandas as pd

# 1. DATA PREPARATION

Prepare the data by performing needed techniques for data cleaning.


## LOAD THE DATASET AND CREATE
### Create from a CSV

In [2]:
df = pd.read_csv('mpe_dataset.csv')

## READ

In [3]:
df.shape

(169, 6)

168 rows and 6 columns

### Show Top 5 and Bottom 5 Rows

In [4]:
df.head(10)

UserID  Day  Duration  Pulse  Maxpulse  Calories
0  202200001  1.0        60    110       130     409.1
1  202200002  2.0        60    117       145     479.0
2  202200003  3.0        60    103       135     340.0
3  202200004  4.0        45    109       175     282.4
4  202200005  5.0        45    117       148     406.0
5  202200006  6.0        60    102       127     300.0
6  202200007  7.0        60    110       136     374.0
7  202200008  8.0        45    104       134     253.3
8  202200009  NaN        30    109       133     195.1
9  202200010  1.0        60     98       124     269.0

### Summary Statistics

In [5]:
df.describe()

UserID         Day    Duration       Pulse    Maxpulse  \
count  1.690000e+02  148.000000  169.000000  169.000000  169.000000   
mean   2.022000e+08    5.453378   63.846154  107.461538  134.047337   
std    3.135624e+01    9.617723   42.299949   14.510259   16.450434   
min    2.022000e+08    0.000000   15.000000   80.000000  100.000000   
25%    2.022000e+08    2.000000   45.000000  100.000000  124.000000   
50%    2.022000e+08    4.000000   60.000000  105.000000  131.000000   
75%    2.022001e+08    6.250000   60.000000  111.000000  141.000000   
max    2.022001e+08   77.000000  300.000000  159.000000  184.000000   

          Calories  
count   164.000000  
mean    375.790244  
std     266.379919  
min      50.300000  
25%     250.925000  
50%     318.600000  
75%     387.600000  
max    1860.400000

### Show Columns and Data Type

In [7]:
df.columns

Index(['UserID', 'Day', 'Duration', 'Pulse', 'Maxpulse', 'Calories'], dtype='object')

In [8]:
df.dtypes

UserID        int64
Day         float64
Duration      int64
Pulse         int64
Maxpulse      int64
Calories    float64
dtype: object

### Print entire DataFrame

In [9]:
df

UserID  Day  Duration  Pulse  Maxpulse  Calories
0    202200001  1.0        60    110       130     409.1
1    202200002  2.0        60    117       145     479.0
2    202200003  3.0        60    103       135     340.0
3    202200004  4.0        45    109       175     282.4
4    202200005  5.0        45    117       148     406.0
..         ...  ...       ...    ...       ...       ...
164  202200083  8.0        60    105       140     290.8
165  202200084  7.0        60    110       145     300.0
166  202200085  6.0        60    115       145     310.2
167  202200086  NaN        75    120       150     320.4
168  202200087  0.0        75    125       150     330.4

[169 rows x 6 columns]

## Cleaning Empty Cells


### Return a new DataFrame with no empty cells:

In [10]:
new_df = df.dropna()
new_df.to_string()

'        UserID   Day  Duration  Pulse  Maxpulse  Calories\n0    202200001   1.0        60    110       130     409.1\n1    202200002   2.0        60    117       145     479.0\n2    202200003   3.0        60    103       135     340.0\n3    202200004   4.0        45    109       175     282.4\n4    202200005   5.0        45    117       148     406.0\n5    202200006   6.0        60    102       127     300.0\n6    202200007   7.0        60    110       136     374.0\n7    202200008   8.0        45    104       134     253.3\n9    202200010   1.0        60     98       124     269.0\n10   202200011   1.0        60    103       147     329.3\n11   202200012   1.0        60    100       120     250.7\n12   202200013   1.0        60    106       128     345.3\n13   202200014   2.0        60    104       132     379.3\n14   202200015   2.0        60     98       123     275.0\n15   202200016   2.0        60     98       120     215.2\n16   202200017   2.0        60    100       120     300

### Print the new dataset

In [11]:
new_df

UserID  Day  Duration  Pulse  Maxpulse  Calories
0    202200001  1.0        60    110       130     409.1
1    202200002  2.0        60    117       145     479.0
2    202200003  3.0        60    103       135     340.0
3    202200004  4.0        45    109       175     282.4
4    202200005  5.0        45    117       148     406.0
..         ...  ...       ...    ...       ...       ...
163  202200082  5.0        45    100       140     280.9
164  202200083  8.0        60    105       140     290.8
165  202200084  7.0        60    110       145     300.0
166  202200085  6.0        60    115       145     310.2
168  202200087  0.0        75    125       150     330.4

[143 rows x 6 columns]

**Note:** By default, the dropna() method returns a new DataFrame, and will not change the `df` (original data set).

Question: What happened to the old data set? Explain your answer and print the new data set.

**Answer: The old data set is still there, and the new_df (new data set) has no empty cells, dropping the rows that had empty cells.**

### Change the original DataFrame using the `inplace = True` argument to remove all rows with NULL values

In [26]:
df = pd.read_csv('mpe_dataset.csv')

df.dropna(inplace = True)
df.to_string()

'        UserID   Day  Duration  Pulse  Maxpulse  Calories\n0    202200001   1.0        60    110       130     409.1\n1    202200002   2.0        60    117       145     479.0\n2    202200003   3.0        60    103       135     340.0\n3    202200004   4.0        45    109       175     282.4\n4    202200005   5.0        45    117       148     406.0\n5    202200006   6.0        60    102       127     300.0\n6    202200007   7.0        60    110       136     374.0\n7    202200008   8.0        45    104       134     253.3\n9    202200010   1.0        60     98       124     269.0\n10   202200011   1.0        60    103       147     329.3\n11   202200012   1.0        60    100       120     250.7\n12   202200013   1.0        60    106       128     345.3\n13   202200014   2.0        60    104       132     379.3\n14   202200015   2.0        60     98       123     275.0\n15   202200016   2.0        60     98       120     215.2\n16   202200017   2.0        60    100       120     300

Print the updated DataFrame

In [27]:
df

UserID   Day  Duration  Pulse  Maxpulse  Calories
0    202200001   1.0        60    110       130     409.1
1    202200002   2.0        60    117       145     479.0
2    202200003   3.0        60    103       135     340.0
3    202200004   4.0        45    109       175     282.4
4    202200005   5.0        45    117       148     406.0
5    202200006   6.0        60    102       127     300.0
6    202200007   7.0        60    110       136     374.0
7    202200008   8.0        45    104       134     253.3
9    202200010   1.0        60     98       124     269.0
10   202200011   1.0        60    103       147     329.3
11   202200012   1.0        60    100       120     250.7
12   202200013   1.0        60    106       128     345.3
13   202200014   2.0        60    104       132     379.3
14   202200015   2.0        60     98       123     275.0
15   202200016   2.0        60     98       120     215.2
16   202200017   2.0        60    100       120     300.0
18   202200019   3.0        60    103       123     323.0
19   202200020   3.0        45     97       125     243.0
20   202200021   3.0        60    108       131     364.2
21   202200022   4.0        45    100       119     282.0
22   202200023   4.0        60    130       101     300.0
23   202200024   4.0        45    105       132     246.0
24   202200025   4.0        60    102       126     334.5
25   202200026   5.0        60    100       120     250.0
26   202200027   5.0        60     92       118     241.0
28   202200029   5.0        60    100       132     280.0
29   202200030   6.0        60    102       129     380.3
30   202200001   6.0        60     92       115     243.0
31   202200001   6.0        45     90       112     180.1
32   202200001   6.0        60    101       124     299.0
33   202200001   7.0        60     93       113     223.0
34   202200001   7.0        60    107       136     361.0
35   202200001   7.0        60    114       140     415.0
36   202200001   7.0        60    102       127     300.0
47   202200040   0.0        45    111       136     300.0
48   202200041   0.0        45    108       129     298.0
49   202200042   0.0        60    111       139     397.6
50   202200043   0.0        60    107       136     380.2
51   202200044   0.0        80    123       146     643.1
52   202200045   1.0        60    106       130     263.0
53   202200046   2.0        60    118       151     486.0
54   202200047   3.0        30    136       175     238.0
55   202200048   4.0        60    121       146     450.7
56   202200049   5.0        60    118       121     413.0
57   202200050   6.0        45    115       144     305.0
58   202200051   7.0        20    153       172     226.4
59   202200052   7.0        45    123       152     321.0
60   202200002   6.0       210    108       160    1376.0
61   202200003   5.0       160    110       137    1034.4
62   202200004   4.0       160    109       135     853.0
63   202200005   3.0        45    118       141     341.0
64   202200006   2.0        20    110       130     131.4
65   202200007   1.0       180     90       130     800.4
75   202200017   1.0        90     98       125     563.2
76   202200018   2.0        45    105       134     251.0
77   202200019   3.0        45    110       141     300.0
78   202200020   4.0       120    100       130     500.4
79   202200021   4.0       270    100       131    1729.0
80   202200022   3.0        30    159       182     319.2
81   202200023   2.0        45    149       169     344.0
82   202200024   1.0        30    103       139     151.1
83   202200025   7.0       120    100       130     500.0
84   202200026   6.0        45    100       120     225.3
85   202200027   5.0        30    151       170     300.0
86   202200028   5.0        45    102       136     234.0
87   202200029   6.0       120    100       157    1000.1
88   202200030   7.0        45    129       103     242.0
89   202200053   1.0        20     83       107      50.3
90   2

**Note:** Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.


Question: What happened to the old data set? Explain your answer and print the new data set.

**Answer: The original DataFrame contained NULL values in some rows. These rows were removed when the inplace = True argument was used. This left the DataFrame with only rows that did not contain NULL values.**

In [28]:
df.shape

(143, 6)

After dropping the rows with empty cells, the data set has 143 rows and 6 columns. 

## Removing Duplicates

*Duplicate rows* are rows that have been registered **more than one time**.
By taking a look at our test data set, we can assume that the `id` column is a unique identifier for each row, and thus, each `id` should only appear once.

To discover duplicates, we can use the `duplicated()` method.

The `duplicated()` method returns a Boolean values for each row:

In [29]:
df

UserID   Day  Duration  Pulse  Maxpulse  Calories
0    202200001   1.0        60    110       130     409.1
1    202200002   2.0        60    117       145     479.0
2    202200003   3.0        60    103       135     340.0
3    202200004   4.0        45    109       175     282.4
4    202200005   5.0        45    117       148     406.0
5    202200006   6.0        60    102       127     300.0
6    202200007   7.0        60    110       136     374.0
7    202200008   8.0        45    104       134     253.3
9    202200010   1.0        60     98       124     269.0
10   202200011   1.0        60    103       147     329.3
11   202200012   1.0        60    100       120     250.7
12   202200013   1.0        60    106       128     345.3
13   202200014   2.0        60    104       132     379.3
14   202200015   2.0        60     98       123     275.0
15   202200016   2.0        60     98       120     215.2
16   202200017   2.0        60    100       120     300.0
18   202200019   3.0        60    103       123     323.0
19   202200020   3.0        45     97       125     243.0
20   202200021   3.0        60    108       131     364.2
21   202200022   4.0        45    100       119     282.0
22   202200023   4.0        60    130       101     300.0
23   202200024   4.0        45    105       132     246.0
24   202200025   4.0        60    102       126     334.5
25   202200026   5.0        60    100       120     250.0
26   202200027   5.0        60     92       118     241.0
28   202200029   5.0        60    100       132     280.0
29   202200030   6.0        60    102       129     380.3
30   202200001   6.0        60     92       115     243.0
31   202200001   6.0        45     90       112     180.1
32   202200001   6.0        60    101       124     299.0
33   202200001   7.0        60     93       113     223.0
34   202200001   7.0        60    107       136     361.0
35   202200001   7.0        60    114       140     415.0
36   202200001   7.0        60    102       127     300.0
47   202200040   0.0        45    111       136     300.0
48   202200041   0.0        45    108       129     298.0
49   202200042   0.0        60    111       139     397.6
50   202200043   0.0        60    107       136     380.2
51   202200044   0.0        80    123       146     643.1
52   202200045   1.0        60    106       130     263.0
53   202200046   2.0        60    118       151     486.0
54   202200047   3.0        30    136       175     238.0
55   202200048   4.0        60    121       146     450.7
56   202200049   5.0        60    118       121     413.0
57   202200050   6.0        45    115       144     305.0
58   202200051   7.0        20    153       172     226.4
59   202200052   7.0        45    123       152     321.0
60   202200002   6.0       210    108       160    1376.0
61   202200003   5.0       160    110       137    1034.4
62   202200004   4.0       160    109       135     853.0
63   202200005   3.0        45    118       141     341.0
64   202200006   2.0        20    110       130     131.4
65   202200007   1.0       180     90       130     800.4
75   202200017   1.0        90     98       125     563.2
76   202200018   2.0        45    105       134     251.0
77   202200019   3.0        45    110       141     300.0
78   202200020   4.0       120    100       130     500.4
79   202200021   4.0       270    100       131    1729.0
80   202200022   3.0        30    159       182     319.2
81   202200023   2.0        45    149       169     344.0
82   202200024   1.0        30    103       139     151.1
83   202200025   7.0       120    100       130     500.0
84   202200026   6.0        45    100       120     225.3
85   202200027   5.0        30    151       170     300.0
86   202200028   5.0        45    102       136     234.0
87   202200029   6.0       120    100       157    1000.1
88   202200030   7.0        45    129       103     242.0
89   202200053   1.0        20     83       107      50.3
90   2

### Check for Duplicates

In [30]:
# Returns True for every row that is a duplicate, otherwise False:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
28     False
29     False
30     False
31     False
32     False
33     False
34     False
35     False
36     False
47     False
48     False
49     False
50     False
51     False
52     False
53     False
54     False
55     False
56     False
57     False
58     False
59     False
60     False
61     False
62     False
63     False
64     False
65     False
75     False
76     False
77     False
78     False
79     False
80     False
81     False
82     False
83     False
84     False
85     False
86     False
87     False
88     False
89     False
90     False
92     False
93     False
94     False
95     False
96     False
97     False
98     False
99     False

Question: What happened to the old data set? Explain your answer.

**Answer: It showed boolean results for every index.**

### Remove Duplicates in `id` column


To remove duplicates, use the drop_duplicates() method.

In [35]:
# Remove duplicates in UserID:
df.drop_duplicates(subset = 'UserID', keep = 'first', inplace = True)

**Note:** The `(inplace = True)` will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.


### Check if there are duplicates again

In [36]:
df

UserID   Day  Duration  Pulse  Maxpulse  Calories
0    202200001   1.0        60    110       130     409.1
1    202200002   2.0        60    117       145     479.0
2    202200003   3.0        60    103       135     340.0
3    202200004   4.0        45    109       175     282.4
4    202200005   5.0        45    117       148     406.0
5    202200006   6.0        60    102       127     300.0
6    202200007   7.0        60    110       136     374.0
7    202200008   8.0        45    104       134     253.3
9    202200010   1.0        60     98       124     269.0
10   202200011   1.0        60    103       147     329.3
11   202200012   1.0        60    100       120     250.7
12   202200013   1.0        60    106       128     345.3
13   202200014   2.0        60    104       132     379.3
14   202200015   2.0        60     98       123     275.0
15   202200016   2.0        60     98       120     215.2
16   202200017   2.0        60    100       120     300.0
18   202200019   3.0        60    103       123     323.0
19   202200020   3.0        45     97       125     243.0
20   202200021   3.0        60    108       131     364.2
21   202200022   4.0        45    100       119     282.0
22   202200023   4.0        60    130       101     300.0
23   202200024   4.0        45    105       132     246.0
24   202200025   4.0        60    102       126     334.5
25   202200026   5.0        60    100       120     250.0
26   202200027   5.0        60     92       118     241.0
28   202200029   5.0        60    100       132     280.0
29   202200030   6.0        60    102       129     380.3
47   202200040   0.0        45    111       136     300.0
48   202200041   0.0        45    108       129     298.0
49   202200042   0.0        60    111       139     397.6
50   202200043   0.0        60    107       136     380.2
51   202200044   0.0        80    123       146     643.1
52   202200045   1.0        60    106       130     263.0
53   202200046   2.0        60    118       151     486.0
54   202200047   3.0        30    136       175     238.0
55   202200048   4.0        60    121       146     450.7
56   202200049   5.0        60    118       121     413.0
57   202200050   6.0        45    115       144     305.0
58   202200051   7.0        20    153       172     226.4
59   202200052   7.0        45    123       152     321.0
76   202200018   2.0        45    105       134     251.0
86   202200028   5.0        45    102       136     234.0
89   202200053   1.0        20     83       107      50.3
90   202200054   1.0       180    101       127     600.1
92   202200056   1.0        30     90       107     105.3
93   202200057   1.0        15     80       100      50.5
94   202200058   1.0        20    150       171     127.4
95   202200059   2.0        20    151       168     229.4
96   202200060   2.0        30     95       128     128.2
97   202200061   2.0        25    152       168     244.2
98   202200062   2.0        30    109       131     188.2
99   202200063   2.0        90     93       124     604.1
100  202200064  22.0        20     95       112      77.7
101  202200065  33.0        90     90       110     500.0
102  202200066   0.1        90     90       100     500.0
103  202200067   2.0        90     90       100     500.4
104  202200068   3.0        30     92       108      92.7
105  202200069   4.0        30     93       128     124.0
106  202200070   5.0       180     90       120     800.3
107  202200071   6.0        30     90       120      86.2
108  202200072  66.0        90     90       120     500.3
109  202200073  77.0       210    137       184    1860.4
110  202200074   5.0        60    102       124     325.2
111  202200075   4.0        45    107       124     275.0
112  202200076   4.0        15    124       139     124.2
113  202200077   1.0        45    100       120     225.3
114  202200078   1.0        60    108       131     367.6
115  202200079   1.0        60    108       151     351.7
116  2

In [34]:
df.shape

(93, 6)

In [37]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
28     False
29     False
47     False
48     False
49     False
50     False
51     False
52     False
53     False
54     False
55     False
56     False
57     False
58     False
59     False
76     False
86     False
89     False
90     False
92     False
93     False
94     False
95     False
96     False
97     False
98     False
99     False
100    False
101    False
102    False
103    False
104    False
105    False
106    False
107    False
108    False
109    False
110    False
111    False
112    False
113    False
114    False
115    False
116    False
117    False
119    False
120    False
121    False
122    False
123    False
124    False
125    False

Question: What happened to the old data set? Explain your answer.


**Answer: We have removed the duplicates, and the `duplicated()` method now returns `False` for every row.**

# END OF 1. DATA PREPARATION

Question: Discuss the techniques you used for data preparation. Explain your answer.

**Answer: 
First, I did some Read and Show operations to get a better understanding of the data set using describe, shape, head to gain summary statistics. Then, I used the `dropna()` method to remove rows that had empty cells. I also used the `drop_duplicates()` method to remove rows that had duplicate `id` values.**

# 2. DATA CORRELATION

Identify the good and bad relationships in the given data set.


## FINDING GOOD RELATIONSHIPS

### A. To show relationship between columns

In [42]:
# For showing relationship of all the columns

df.corr()

UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.147270  0.010627  0.088296  0.081702  0.040039
Day       0.147270  1.000000  0.441471  0.046026  0.139963  0.591747
Duration  0.010627  0.441471  1.000000 -0.169937 -0.052151  0.856017
Pulse     0.088296  0.046026 -0.169937  1.000000  0.803504  0.147037
Maxpulse  0.081702  0.139963 -0.052151  0.803504  1.000000  0.254320
Calories  0.040039  0.591747  0.856017  0.147037  0.254320  1.000000

In [43]:
# For showing relationship between two specific columns
df["Duration"].corr(df["Calories"])

0.8560173544756272

### B. Add a parameter method in corr() function

In [44]:
# pearson: standard correlation coefficient 
df.corr('pearson')

UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.147270  0.010627  0.088296  0.081702  0.040039
Day       0.147270  1.000000  0.441471  0.046026  0.139963  0.591747
Duration  0.010627  0.441471  1.000000 -0.169937 -0.052151  0.856017
Pulse     0.088296  0.046026 -0.169937  1.000000  0.803504  0.147037
Maxpulse  0.081702  0.139963 -0.052151  0.803504  1.000000  0.254320
Calories  0.040039  0.591747  0.856017  0.147037  0.254320  1.000000

In [45]:
# spearman: Spearman rank correlation
df.corr('spearman')

UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.201422 -0.047843  0.069412  0.102369  0.015973
Day       0.201422  1.000000  0.002586  0.018790  0.049266 -0.005735
Duration -0.047843  0.002586  1.000000 -0.208068 -0.206392  0.816602
Pulse     0.069412  0.018790 -0.208068  1.000000  0.814217  0.160723
Maxpulse  0.102369  0.049266 -0.206392  0.814217  1.000000  0.149637
Calories  0.015973 -0.005735  0.816602  0.160723  0.149637  1.000000

In [46]:
# kendall: Kendall Tau correlation coefficient 
df.corr('kendall')

UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.157043 -0.042944  0.055739  0.074308 -0.004679
Day       0.157043  1.000000  0.004405  0.009241  0.040868 -0.001971
Duration -0.042944  0.004405  1.000000 -0.167793 -0.166282  0.688311
Pulse     0.055739  0.009241 -0.167793  1.000000  0.666828  0.169037
Maxpulse  0.074308  0.040868 -0.166282  0.666828  1.000000  0.117487
Calories -0.004679 -0.001971  0.688311  0.169037  0.117487  1.000000

### 1. Find and identify data with good correlation.

Using Pearson correlation as default

In [47]:
# Find and identify data with good correlation.

# Use any parameter method to find good correlation
df.corr(method='pearson')


UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.147270  0.010627  0.088296  0.081702  0.040039
Day       0.147270  1.000000  0.441471  0.046026  0.139963  0.591747
Duration  0.010627  0.441471  1.000000 -0.169937 -0.052151  0.856017
Pulse     0.088296  0.046026 -0.169937  1.000000  0.803504  0.147037
Maxpulse  0.081702  0.139963 -0.052151  0.803504  1.000000  0.254320
Calories  0.040039  0.591747  0.856017  0.147037  0.254320  1.000000

**What are the data with good correlation?**

In [48]:
# Show relationship between two or more specific columns that show good correlation
df["Duration"].corr(df["Calories"])

0.8560173544756272

**Duration and Calories Burned have a good correlation because the value is close to 1.**

### 2. Find and identify data with bad correlation.


Using Pearson correlation as default

In [49]:
df.corr(method='pearson')

UserID       Day  Duration     Pulse  Maxpulse  Calories
UserID    1.000000  0.147270  0.010627  0.088296  0.081702  0.040039
Day       0.147270  1.000000  0.441471  0.046026  0.139963  0.591747
Duration  0.010627  0.441471  1.000000 -0.169937 -0.052151  0.856017
Pulse     0.088296  0.046026 -0.169937  1.000000  0.803504  0.147037
Maxpulse  0.081702  0.139963 -0.052151  0.803504  1.000000  0.254320
Calories  0.040039  0.591747  0.856017  0.147037  0.254320  1.000000

**What are the data with bad correlation?**

In [50]:
# Show relationship between two or more specific columns that show bad correlation
df["Duration"].corr(df["Maxpulse"])

-0.05215085188731025

**Duration and Maxpulse have a bad correlation because the value is not just close to 0, but it is negative!**

In [52]:
# Another data correlation that show bad correlation
df["Duration"].corr(df["Pulse"])

-0.16993724319469103

Question: Discuss the techniques you used for data correlation.

**Answer: I used the `corr()` function to find the correlation between columns. I used the default Pearson correlation method. It's also possible to gain more but small insights when using the `sort_values()` function to sort the values in ascending order. This way, we can see the columns with the highest correlation. Determining the good and bad relationships is a matter of interpretation that's why I used the default Pearson correlation method.**

# END OF 2. DATA CORRELATION